### ISSUE TITLE
Trying to backward through the graph a second time, but the buffers have already been freed. Specify retain_graph=True when calling backward the first time.

### ISSUE THREAD

**Issue by shaoeric** [chxy95/Deep-Mutual-Learning#5]
when updating the sub network, is there any need to retain graph like
`loss.backward(retain_graph=True)`
because when i reproduce the procedure, the code runs wrong, but i dont know if retaining the graph is a correct operation

**Comment by chxy95** [comment]
@shaoeric When I ran this code, this error didn't appear. But it seems like to be correct if your code does report this error. There may indeed be some bugs in this code because it is not under maintenance. The core of the code is the implementation of KL divergence loss and I verified that this part is correct.

**Comment by shaoeric** [comment]
@chxy95 So sorry, i solved the problem without closing the issue.  As for this problem, i have got some solutions. 
1. First, check out pytorch version, perhaps a newer version causes the confusing problem
2. check whether the teacher model's output is detached from the graph `teacher_out = teacher_out.detach()`
3. if we have more than one teacher models, putting their outputs into a list is better than a tensor, even if the tensor is detached.

**Comment by KeanuWongves** [comment]
Yeah, u should use detach , because your gridient will be freed the first time you use kl and backward

**Comment by siddhawan** [comment]
> @chxy95 So sorry, i solved the problem without closing the issue. As for this problem, i have got some solutions.
> 
> 1. First, check out pytorch version, perhaps a newer version causes the confusing problem
> 2. check whether the teacher model's output is detached from the graph `teacher_out = teacher_out.detach()`
> 3. if we have more than one teacher models, putting their outputs into a list is better than a tensor, even if the tensor is detached.

Hi, can you please help me with solution. I have 3 models same as stated in the original paper i was also getting the same error then I read you comment and used detach() on all the outputs of the 3 models and in inner loop where the loss is calculated for each model I put .requires_grad_() and calculate loss.backwarsd() but loss is None. I suppose the computational graph is disconnected after the detach(). 

So I transferred the  outputs(logits) of all models in python list and the logic is same as its in the code on github. But still it gives error you stated above.
Can you please help with the changes you have made 🙏
